### PR TITLE
inititial version

### PR DESCRIPTION
**Pull request by aklein13**

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -0,0 +1,2 @@
+requests
+test
```
Comment (aklein13) [inline requirements.txt]: ups

Comment (aklein13) [reply requirements.txt]: i will remove it in later commits

File: .gitignore [hunk 2]
```diff
@@ -99,3 +99,5 @@ ENV/
 
 # mypy
 .mypy_cache/
+
+cache/
```
Comment (PythonicNinja) [inline .gitignore]: new line

File: olx.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    html_parser = BeautifulSoup(offer_markup, "html.parser")
+    url = html_parser.find(class_="linkWithHash").attrs['href']
+    if not url:
+        # detail url is not present
+        return []
+    if urlparse(url).hostname not in WHITELISTED_DOMAINS:
+        # domain is not supported by this backend
+        return []
+    return url
+    # return {
```
Comment (PythonicNinja) [inline olx.py]: removal of commented code

File: olx.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+def get_title(offer_markup):
+    html_parser = BeautifulSoup(offer_markup, "html.parser")
+    title = html_parser.h1.contents
+    return title[0].replace("\n", "").replace("  ","")
+
+
+def get_price(offer_markup):
+    html_parser = BeautifulSoup(offer_markup, "html.parser")
+    price = html_parser.find(class_="xxxx-large").contents
```
Comment (PythonicNinja) [inline olx.py]: maybe .text

File: olx.py [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+def get_phone(offer_markup):
+    html_parser = BeautifulSoup(offer_markup, "html.parser")
+    phone = html_parser.find(class_="xx-large").contents
+    return phone[0]
+
+
+def parse_avalible_offers(markup):
+    html_parser = BeautifulSoup(markup, "html.parser")
+    offers = html_parser.find_all(class_='offer')
+    parsed_offers = [parse_offer(str(offer)) for offer in offers if offer]
```
Comment (PythonicNinja) [inline olx.py]: `if offer and offer is not featured` then last lines can be removed

File: olx.py [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+    # delete featured offers (always first 3)
+    for i in range(3):
+        del parsed_offers[0]
+    return parsed_offers
+
+
+def parse_description(markup,url):
+    html_parser = BeautifulSoup(markup, "html.parser")
+    descriptions = html_parser.find_all(class_='offerbody')
```
Comment (PythonicNinja) [inline olx.py]: it has more than one?

File: olx.py [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+
+
+def get_category(main_category, subcategory, detail_category, region, **filters):
+    parsed_content = []
+    page = 1
+    page_attr = None
+    while True:
+        url = get_url(main_category, subcategory, detail_category, region, page_attr, **filters)
+        response = get_content_for_url(url)
+        print(url)
```
Comment (PythonicNinja) [inline olx.py]: removal of print, maybe use `log.info`

File: olx.py [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+
+
+def get_description(parsed_urls):
+    i = 0
+    descriptions = []
+    for url in parsed_urls:
+        response = get_content_for_url(url)
+        descriptions.append(parse_description(response.content,url))
+        i+=1
+        if i>5:
```
Comment (PythonicNinja) [inline olx.py]: why i>5?

File: olx.py [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+def get_price(offer_markup):
+    html_parser = BeautifulSoup(offer_markup, "html.parser")
+    price = html_parser.find(class_="xxxx-large").contents
+    return price[0]
+
+
+def get_phone(offer_markup):
+    html_parser = BeautifulSoup(offer_markup, "html.parser")
+    phone = html_parser.find(class_="xx-large").contents
+    return phone[0]
```
Comment (PythonicNinja) [inline olx.py]: maybe text?

File: olx.py [hunk 10]
```diff
@@ -26,21 +26,52 @@ def flatten(container):
         else:
             yield i
 
-# TODO: remove polish chars
-def city_name(input):
+
+def city_name(city):
+    POLISH_CHARACTERS = {"ą": "a", "ć": "c", "ę": "e", "ł": "l", "ń": "n", "ó": "o", "ś": "s", "ż": "z", "ź": "z"}
```
Comment (PythonicNinja) [inline olx.py]: maybe move it to global scope?

File: olx.py [hunk 11]
```diff
@@ -26,21 +26,52 @@ def flatten(container):
...
     out = ""
-    for char in input:
+    city = city.lower()
+    for char in city:
```
Comment (PythonicNinja) [inline olx.py]: `for char in city.lower():`

Comment (PythonicNinja) [reply olx.py]:
```
+def replace_all(text, dic):
 +    for i, j in dic.items():
 +        text = text.replace(i, j)
 +    return text
```

`replace_all(city.lower(), POLISH_CHARACTERS_MAPPING)`

File: olx.py [hunk 12]
```diff
@@ -26,21 +26,52 @@ def flatten(container):
...
+
+def url_furniture(furniture):
+    if furniture:
+        return "search%5Bfilter_enum_furniture%5D%5B0%5D=yes"
+    else:
+        return "search%5Bfilter_enum_furniture%5D%5B0%5D=no"
+
+
+def url_floor(floor):
+    return "search%5Bfilter_enum_floor_select%5D%5B0%5D=floor_" + str(floor)
```
Comment (PythonicNinja) [inline olx.py]: maybe use format here?

File: olx.py [hunk 13]
```diff
@@ -49,25 +80,26 @@ def get_url(page=None, *args):
...
 
+
+# TODO: Caching for long urls
 @caching
 def get_content_for_url(url):
     response = requests.get(url, allow_redirects=False)
     try:
         response.raise_for_status()
-    except requests.HTTPError as e:
+    except requests.HTTPError:
```
Comment (PythonicNinja) [inline olx.py]: you shouldn't cache httperrors?

File: .gitignore (hunk 2 above)
Comment (PythonicNinja) [reply .gitignore]: ?

File: olx.py [hunk 14]
```diff
@@ -0,0 +1,342 @@
...
+                    url += "?" + page + "&" + element + "&"
+                else:
+                    url += "?" + element + "&"
+            else:
+                url += element + "/"
+    return url
+
+
+def url_buildtype(type):
+    return "search%5Bfilter_enum_builttype%5D%5B0%5D=" + type
```
Comment (PythonicNinja) [inline olx.py]: type is builtin, maybe consider `type_`

Comment (PythonicNinja) [reply olx.py]: or change name

File: olx.py [hunk 15]
```diff
@@ -0,0 +1,342 @@
...
+            break
+        print("Loaded page {} of offers".format(page))
+        offers = parse_available_offers(response.content)
+        if offers is None:
+            break
+        parsed_content.append(offers)
+        page += 1
+        page_attr = "page={}".format(page)
+    parsed_content = list(flatten(parsed_content))
+    print("Loaded " + str(len(parsed_content)) + " offers")
```
Comment (PythonicNinja) [inline olx.py]: we should remove prints and maybe use logging[1]


[1] - https://docs.python.org/3/howto/logging.html

File: olx.py [hunk 16]
```diff
@@ -0,0 +1,342 @@
...
+def get_description(parsed_urls):
+    i = 0
+    descriptions = []
+    for url in parsed_urls:
+        response = get_content_for_url(url)
+        try:
+            descriptions.append(parse_offer(response.content, url))
+        except AttributeError:
+            print("This offer is not available anymore")
+        if DEBUG:
```
Comment (PythonicNinja) [inline olx.py]: var `i` doesn't tell anything

File: olx.py [hunk 17]
```diff
@@ -0,0 +1,342 @@
...
+            i += 1
```
Comment (PythonicNinja) [inline olx.py]: it shouldn't be in such a way

File: olx.py [hunk 18]
```diff
@@ -0,0 +1,342 @@
...
+        output.append(element.attrs["src"])
+
+    return output
+
+
+def get_date_added(offer_markup):
+    html_parser = BeautifulSoup(offer_markup, "html.parser")
+    date = html_parser.find(class_="offer-titlebox__details").em.contents
+    try:
+        return date[4].replace("Dodane", "").replace("\n", "").replace("  ", "").replace("o ", "").replace(", ", "")
```
Comment (PythonicNinja) [inline olx.py]: looks like not dry to me

File: olx.py [hunk 19]
```diff
@@ -0,0 +1,342 @@
...
+    for char in price:
+        if char.isdigit():
+            output += char
+    return int(output)
+
+
+def get_yardage(offer_markup):
+    html_parser = BeautifulSoup(offer_markup, "html.parser")
+    try:
+        yardage = html_parser.sup.parent.text
```
Comment (PythonicNinja) [inline olx.py]: this will not raise value error

File: olx.py (hunk 10 above)
Comment (PLJohnny) [reply olx.py]: You could do `unicodedata.normalize('NFKD', city.lower()).encode('ascii', 'ignore').decode("utf-8")` - it gives you a lowercase string stripped of diacritics

File: olx.py (hunk 19 above)
Comment (aklein13) [reply olx.py]: It will raise exception that there is no attribute parent for NoneType (sup) if there is no m^2 symbol in offer

Comment (aklein13) [reply olx.py]: Nvm fixed it to AttributeError

File: olx.py [hunk 20]
```diff
@@ -0,0 +1,386 @@
...
+import requests
+from bs4 import BeautifulSoup
+
+# own modules
+from scrapper_helpers.utils import caching
+
+BASE_URL = 'https://www.olx.pl/'
+OFFERS_FEATURED_PER_PAGE = 3
+POLISH_CHARACTERS_MAPPING = {"ą": "a", "ć": "c", "ę": "e", "ł": "l", "ń": "n", "ó": "o", "ś": "s", "ż": "z", "ź": "z"}
+DEBUG = True
```
Comment (PythonicNinja) [inline olx.py]: at the very top:
`import os`
and then here:
`DEBUG = os.environ.get('DEBUG')`

File: olx.py [hunk 21]
```diff
@@ -0,0 +1,386 @@
...
+    return "search%5Bfilter_float_price%3Afrom%5D={0}".format(str(price))
+
+
+def url_price_to(price):
+    return "search%5Bfilter_float_price%3Ato%5D={0}".format(str(price))
+
+
+def url_rooms(number):
+    # 4 and more rooms as 4
+    if number > 4:
```
Comment (PythonicNinja) [inline olx.py]: remove it

File: olx.py [hunk 22]
```diff
@@ -36,55 +43,148 @@ def flatten(container):
...
+    It also should replace white spaces with dashes.
+
+    :param city: City name not in olx url format
+    :type city: str
+    :return: Valid olx url city name
+    :rtype: str
+
+    :Example:
+
+    city_name("Ruda Śląska") >> "ruda-slaska"
```
Comment (PythonicNinja) [inline olx.py]:
```
>> city_name("Ruda Śląska")
"ruda-slaska"
```

File: olx.py [hunk 23]
```diff
@@ -0,0 +1,519 @@
...
+    :type number: int
+    :return: Olx url search filer for number of rooms
+    :rtype: str
+    """
+    numbers = {1: "one", 2: "two", 3: "three", 4: "four"}
+    return "search%5Bfilter_enum_rooms%5D%5B0%5D={0}".format(numbers.get(number, "four"))
+
+
+def url_surface_from(minimum):
+    """ Generates olx search filter for minimal surface
```
Comment (PythonicNinja) [inline olx.py]: olx should be probably mentioned in one format

File: olx.py [hunk 24]
```diff
@@ -0,0 +1,519 @@
...
+    :param floor: Desired offer floor
+    :type floor: int
+    :return: Olx url search filter for floor
+    :rtype: str
+    """
+    floor_id = 11 if floor > 10 and floor != 17 else str(floor)
+    return "search%5Bfilter_enum_floor_select%5D%5B0%5D=floor_{0}".format(floor_id)
+
+
+def url_builttype(builttype):
```
Comment (PythonicNinja) [inline olx.py]: `built_type`

File: olx.py [hunk 25]
```diff
@@ -0,0 +1,519 @@
...
+
+    :param main_category: Main category
+    :param subcategory: Sub category
+    :param detail_category: Detail category
+    :param region: Region
+    :param args: Additional search filters
+    :return: List of all offer urls for category
+    :rtype: list
+    """
+    parsed_content = []
```
Comment (PythonicNinja) [inline olx.py]: probably you can do that in the very same way:
https://github.com/limebrains/pyotodom/pull/1/files#diff-c9b9cd160564be25826d28dccb5572a8R132

File: tests.py [hunk 26]
```diff
@@ -0,0 +1,162 @@
...
+def test_flatten(c):
+    result = olx.flatten(c)
+    for element in result:
+        assert not isinstance(element, list)
+
+
+@pytest.mark.parametrize('d', [
+    (range(3200)),
+])
+def test_get_paraments(d):
```
Comment (PythonicNinja) [inline tests.py]: i would also test how it creates url with both `from` and `to`

File: tests.py [hunk 27]
```diff
@@ -0,0 +1,162 @@
...
+    assert olx.url_price_from(d) == "search%5Bfilter_float_price%3Afrom%5D={0}".format(str(d))
+    assert olx.url_price_to(d) == "search%5Bfilter_float_price%3Ato%5D={0}".format(str(d))
+    assert olx.url_surface_from(d) == "search%5Bfilter_float_m%3Afrom%5D={0}".format(str(d))
+    assert olx.url_surface_to(d) == "search%5Bfilter_float_m%3Ato%5D={0}".format(str(d))
+
+
+@pytest.mark.parametrize('e', ["blok", "apartamentowiec", "kamienica"])
+def test_builttype(e):
```
Comment (PythonicNinja) [inline tests.py]: i would rename those argument from alphabet names `a,b,c,d...`

File: tests.py [hunk 28]
```diff
@@ -0,0 +1,162 @@
...
+
+
+@pytest.mark.parametrize("l", [
+    str(offer) for offer in offers if offer
+])
+def test_parse_offer_url(l):
+    olx.parse_offer_url(l)
+
+
+response = olx.get_content_for_url(OFFER_URL)
```
Comment (PythonicNinja) [inline tests.py]: it should be fixture

File: tests.py [hunk 29]
```diff
@@ -0,0 +1,162 @@
...
+    for img in images:
+        assert "https://" in img
+
+
+@pytest.mark.parametrize("a", [offer_content])
+def test_get_date_added(a):
+    assert olx.get_date_added(a)
+
+
+offer_content = str(html_parser.body)
```
Comment (PythonicNinja) [inline tests.py]: it should be fixture

**Pull request merged**